### PR TITLE
[RuntimeEnv] exec_worker failing on windows installations with spaces

### LINKED ISSUE #24574: [<Ray component: Core|RLlib|etc...>]
### What happened + What you expected to happen

I just installed Ray with Pycharm. and I tried to run the code as below.
`import ray`
`ray.init()`
`from ray import tune`
`tune.run("PPO",config = {"env": "CartPole-v1",
                         } )`

then I got the error as below
`
2022-05-07 10:14:36,980	INFO trial_runner.py:803 -- starting PPO_CartPole-v1_fbf28_00000
2022-05-07 10:14:37,068	ERROR syncer.py:119 -- Log sync requires rsync to be installed.
(pid=) C:\Users\Shaojie Liu\AppData\Local\Programs\Python\Python37\python.exe: can't open file 'C:\Users\Shaojie': [Errno 2] No such file or directory
(pid=) 2022-05-07 10:14:41,180	INFO context.py:67 -- Exec'ing worker with command: "C:\Users\Shaojie Liu\AppData\Local\Programs\Python\Python37\python.exe" C:\Users\Shaojie Liu\AppData\Roaming\Python\Python37\site-packages\ray\workers/default_worker.py --node-ip-address=127.0.0.1 --node-manager-port=6876 --object-store-name=tcp://127.0.0.1:27962 --raylet-name=tcp://127.0.0.1:60187 --redis-address=None --storage=None --temp-dir=C:\Users\SHAOJI~1\AppData\Local\Temp\ray --metrics-agent-port=13502 --logging-rotate-bytes=536870912 --logging-rotate-backup-count=5 --gcs-address=127.0.0.1:16018 --redis-password=5241590000000000 --startup-token=12 --runtime-env-hash=1728481729
2022-05-07 10:14:56,229	WARNING tune.py:657 -- SIGINT received (e.g. via Ctrl+C), ending Ray Tune run. This will try to checkpoint the experiment state one last time. Press CTRL+C one more time (or send SIGINT/SIGKILL/SIGTERM) to skip. 
`

### Versions / Dependencies

Python version is 3.7, Ray version is 1.12. 

### Reproduction script

`import ray`
`ray.init()`
`from ray import tune`
`tune.run("PPO",config = {"env": "CartPole-v1",
                         } )`

### Issue Severity

_No response_

### PR DESCRIPTION
**Pull request by bushshrub**
## Why are these changes needed?

On Windows systems, the `exec_worker` method may fail due to spaces being present in arguments that are file paths. This addresses said issue.

## Related issue number
Closes #24574

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by mwtian** [comment]
Adding folks familiar with runtime env. Is it possible to add a test for this?

File: python/ray/_private/runtime_env/context.py [hunk 1]
```diff
@@ -49,6 +49,9 @@ def exec_worker(self, passthrough_args: List[str], language: Language):
             executable = f'"{self.py_executable}"'  # Path may contain spaces
         elif language == Language.PYTHON:
             executable = f"exec {self.py_executable}"
+            for i, arg in enumerate(passthrough_args):
```
Comment (mwtian) [inline python/ray/_private/runtime_env/context.py:52]: It looks like under Windows, the above branch will be taken instead of this branch. Does this change fix your issue?

Comment (xiurobert) [reply python/ray/_private/runtime_env/context.py:52]: Ah, I blundered. Must have been too tired when I patched this. I'll fix it in a bit.

**Comment by xiurobert** [comment]
Pinging https://github.com/SJ-L

**Comment by edoakes** [comment]
@xiurobert what's the testing plan here?

**Comment by xiurobert** [comment]
Not 100% sure, but I'm thinking creating a python file with spaces and seeing if this change fixes it? Just need to locate where the runtime_env tests are stored.

**Comment by xiurobert** [comment]
Honestly, I think this code I wrote is more of a hack than actually a real solution to the problem. The real solution should be to either handle arguments that are paths upstream, or inform the method that paths are being passed in.

**Comment by architkulkarni** [comment]
@xiurobert the tests are in the files `test_runtime_env_*.py`, perhaps you could try adding a test there?  Let me know if you have any questions.

The fix looks reasonable to me, @mattip could you please help review this PR and see if there are any issues or edge cases for Windows?

**Comment by xiurobert** [comment]
> @xiurobert the tests are in the files `test_runtime_env_*.py`, perhaps you could try adding a test there? Let me know if you have any questions.
> 
> The fix looks reasonable to me, @mattip could you please help review this PR and see if there are any issues or edge cases for Windows?

Sounds good, I will add in a test.

**Comment by xiurobert** [comment]
@architkulkarni I created the tests under `test_runtime_env_ray_minimal.py`. I had to abstract out the argument fixing for windows into a separate method so I could actually test it.

**Comment by architkulkarni** [comment]
Thanks for adding the test, it looks good!  Let's wait for @mattip's review.

File: python/ray/tests/test_runtime_env_ray_minimal.py [hunk 2]
```diff
@@ -79,6 +80,20 @@ def f():
     assert ray.get(f.remote()) == 1
 
 
+@pytest.mark.skipif(
+    sys.platform != "win32", reason="Windows-specific test"
+)
+def test_exec_worker_parse_correct(shutdown_only):
+    runtime_env_ctx = RuntimeEnvContext()
+    runtime_env_ctx._fix_windows_args(["C:\Program Files", "some", "other", "arguments"])
+    assert runtime_env_ctx.exec_worker_args == [
```
Comment (mattip) [inline python/ray/tests/test_runtime_env_ray_minimal.py:89]:
```suggestion
    ret = runtime_env_ctx._fix_windows_args(["C:\Program Files", "some", "other", "arguments"])
    assert ret == [
```

File: python/ray/_private/runtime_env/context.py [hunk 3]
```diff
@@ -42,11 +42,30 @@ def serialize(self) -> str:
...
+        Note that the function only works if the file path actually exists.
+
+        Args:
+            passthrough_args: the passthrough arguments
+
+        Returns:
+            Fixed passthrough arguments
+        """
+        for i, arg in enumerate(passthrough_args):
+            if os.path.isfile(arg) or os.path.isdir(arg):
```
Comment (mattip) [inline python/ray/_private/runtime_env/context.py:59]: Why not `if os.path.exists(arg)` ?

**Comment by mattip** [comment]
This seems to have broken runtime environment tests

File: python/ray/_private/runtime_env/context.py (hunk 3 above)
Comment (xiurobert) [reply python/ray/_private/runtime_env/context.py]: That makes sense. Updated in the latest commits

**Comment by xiurobert** [comment]
Should be fine now. @architkulkarni

File: python/ray/tests/test_runtime_env_ray_minimal.py [hunk 4]
```diff
@@ -79,6 +80,20 @@ def f():
...
+    runtime_env_ctx = RuntimeEnvContext()
+    ret = runtime_env_ctx._fix_windows_args(["C:\Program Files", "some", "other", "arguments"])
+    assert ret == [
+        '"C:\Program Files"',
+        "some",
+        "other",
+        "arguments",
+    ]
+
+
```
Comment (mattip) [inline python/ray/tests/test_runtime_env_ray_minimal.py:96]: This will fail if the string is already quoted. The easiest way to test this is to pass the argument through `_fix_windows_args` a couple of times. So something like
```
import subprocess
...
    # inside the test, needs linting
    args = ["C:\Program Files"]
    # Check that already-quoted-strings are not requoted
    for i in range(3):
       args = runtime_env_ctx.fix_windows_args(args)
       subprocess.check_call([sys.executable, "-c", f"import os; os.listdir({args[0]})"])
```

Comment (xiurobert) [reply python/ray/tests/test_runtime_env_ray_minimal.py:96]: Let me start my windows VM and test this out

Comment (mattip) [reply python/ray/tests/test_runtime_env_ray_minimal.py:96]: It seems indeed that the added check will pass the current implementation. So it is worth adding the check to ensure future "improvements" do not break.

Comment (xiurobert) [reply python/ray/tests/test_runtime_env_ray_minimal.py:96]: Ok, just checked, if I add double quotes:
```python
import os
assert os.path.exists('"C:\Program Files"') == False
```
That works.

Comment (xiurobert) [reply python/ray/tests/test_runtime_env_ray_minimal.py:96]: Yup, that's probably the best course of action.

Comment (xiurobert) [reply python/ray/tests/test_runtime_env_ray_minimal.py:104]: Latest commits should take care of the issues

**Review by mattip — COMMENTED** [review]
LGTM assuming tests pass

**Comment by architkulkarni** [comment]
```
\\?\C:\Users\ContainerAdministrator\AppData\Local\temp\Bazel.runfiles_j411cim1\runfiles\com_github_ray_project_ray\python\ray\tests\test_runtime_env_ray_minimal.py:88: DeprecationWarning: invalid escape sequence \P
--
  | ["C:\Program Files", "some", "other", "arguments"]
```
Do we just need `"C:\\Program Files"` here?

**Comment by xiurobert** [comment]
Probably, but the other arguments are left there to ensure any future modifications do not try to quote non paths.

**Comment by architkulkarni** [comment]
Ah right, I was just talking about the "invalid escape sequence \P" warning for the first one.

**Comment by architkulkarni** [comment]
It looks like some runtime env tests are timing out. The simplest and easiest to debug might be test_runtime_env_env_vars: https://buildkite.com/ray-project/ray-builders-pr/builds/34212#01812809-198e-4e2a-92e9-f21ee6921bf5/3405-6289 I wonder if you can reproduce this locally?

**Comment by xiurobert** [comment]
Strange, I did not actually make any changes in that test. I'll give it a try later.

**Comment by xiurobert** [comment]
Could not reproduce issue with runtime_env_vars on my machine.

**Comment by architkulkarni** [comment]
Ah so sorry, should have mentioned that it was only failing on Windows, not on Linux.  It's [still failing after a retry](https://buildkite.com/ray-project/ray-builders-pr/builds/34212#01812ab0-228b-455a-a641-64f915e28c0b/3402-5600), so it's probably a legitimate issue stemming from this PR and not just flakiness.  We'd need to fix this before merging.

**Comment by xiurobert** [comment]
Yep, I'm looking into it. Am going to try building on Windows.

**Comment by xiurobert** [comment]
Do we need to make a call to `os.system`?

**Comment by xiurobert** [comment]
An additional note, maybe os.system can be changed out for `subprocess` instead. Could probably get around the whole issue with quoting altogether.

**Comment by mattip** [comment]
The code that calls `os.system` is not far from the code in this PR https://github.com/ray-project/ray/blob/a9d8da0100e680e5fa88c88fe99f139f243c86f8/python/ray/_private/runtime_env/context.py#L71-L79

I think `os.system()` could be changed to `subprocess.check_call()`?

**Comment by xiurobert** [comment]
Yup, I think that would be best. Will take care of that asap.

**Comment by mattip** [comment]
I did the same change in #25546, which also further cleans up the code and also modifies the mock test to use `subprocess.run`.

File: python/ray/tests/test_runtime_env_ray_minimal.py [hunk 5]
```diff
@@ -78,7 +80,6 @@ def f():
 
     assert ray.get(f.remote()) == 1
 
-
```
Comment (mattip) [inline python/ray/tests/test_runtime_env_ray_minimal.py:81]: Can you revert the changes to this file?

File: python/ray/_private/runtime_env/context.py [hunk 6]
```diff
@@ -69,7 +68,7 @@ def exec_worker(self, passthrough_args: List[str], language: Language):
         command_str = " && ".join(self.command_prefix + [exec_command])
         logger.debug(f"Exec'ing worker with command: {command_str}")
         if sys.platform == "win32":
-            os.system(command_str)
+            subprocess.check_call([executable, *passthrough_args])
```
Comment (mattip) [inline python/ray/_private/runtime_env/context.py:71]: Did you try `check_call`? I think some of the processes exit with `1` as an exit code, so in the other PR I used `subprocess.run` instead. I think the error checking is done by other ray mechanisms: checking that the worker registers back with the raylet and emits proper heartbeats.

File: python/ray/tests/test_runtime_env_ray_minimal.py (hunk 5 above)
Comment (xiurobert) [reply python/ray/tests/test_runtime_env_ray_minimal.py:81]: Sure

File: python/ray/_private/runtime_env/context.py (hunk 6 above)
Comment (xiurobert) [reply python/ray/_private/runtime_env/context.py:71]: I should probably switch to subprocess.run yeah, according to the python docs it's better

**Comment by xiurobert** [comment]
Different tests are failing this time.

**Comment by xiurobert** [comment]
It looks like this test is broken

https://buildkite.com/ray-project/ray-builders-pr/builds/34930#0181487a-2a63-4c0c-81ab-60b32536d4b4/5810-5814

**Comment by xiurobert** [comment]
Oh, very sorry about that. I rebased the branch and github did some weird things.

**Comment by xiurobert** [comment]
Sorry, let me close this PR. I will re-open a new one
